### PR TITLE
Precharge Timer creation to prevent early precharge termination

### PR DESCRIPTION
**Pull request by LucasZahlan**
### Describe changes

1. implemented a timer with the minimum time constant to the BMS boss.

### Impact

1. Prevent bms boss precharge termination early by implementing a minimum time before closing contactors.

### Test Plan

- [ ] Test precharge termination.

### CODE REVIEW

File: components/bms_boss/include/Sys.h [hunk 1]
```diff
@@ -55,3 +55,4 @@ bool SYS_SFT_checkBrusaChargerTimeout(void);
 bool SYS_SFT_checkElconChargerTimeout(void);
 void SYS_stopCharging(void);
 void SYS_continueCharging(void);
+void closeSYS_SFT_cycleContacts(void);
```
Comment (JoshLafleur) [inline components/bms_boss/include/Sys.h:58]: This is done by `SYS_SFTY_cycleContacts`

File: components/bms_boss/src/BMS.c [hunk 2]
```diff
@@ -138,12 +143,26 @@ static void BMS100Hz_PRD(void)
                 float32_t chg_voltage = 0.0f;
                 const bool mc_valid = (CANRX_get_signal(VEH, PM100DX_tractiveSystemVoltage, &ts_voltage) == CANRX_MESSAGE_VALID);
                 const bool chg_valid = (CANRX_get_signal(VEH, BRUSA513_dcBusVoltage, &chg_voltage) == CANRX_MESSAGE_VALID);
+                
+                if (SYS.contacts == SYS_CONTACTORS_PRECHARGE && !precharge_timer_started)
+                {
+                    drv_timer_init(&precharge_timer);
```
Comment (JoshLafleur) [inline components/bms_boss/src/BMS.c:149]: TYhis should be done only once in `BMS_init`

File: components/bms_boss/src/BMS.c [hunk 3]
```diff
@@ -138,12 +143,26 @@ static void BMS100Hz_PRD(void)
...
+                    drv_timer_start(&precharge_timer, 0);
```
Comment (JoshLafleur) [inline components/bms_boss/src/BMS.c:150]: The timer should be started above in 
```
if (SYS.contacts == SYS_CONTACTORS_OPEN)
{
    SYS_SFT_cycleContacts();
    <here>
}
```

File: components/bms_boss/src/BMS.c [hunk 4]
```diff
@@ -20,9 +21,13 @@
 #include "FeatureDefines_generated.h"
 
 #define CURRENT_SENSE_V_per_A 0.005f
+#define PRECHARGE_MIN_TIME_MS 1320U
 
 BMSB_S BMS;
 
+static drv_timer_S precharge_timer;
+static bool precharge_timer_started = false;
+
```
Comment (JoshLafleur) [inline components/bms_boss/src/BMS.c:30]: You shouldnt need a precharge timer started. You know the timer hasnt started because it should be in a `STOPPED` state

File: components/bms_boss/src/BMS.c [hunk 5]
```diff
@@ -138,12 +143,26 @@ static void BMS100Hz_PRD(void)
                 float32_t chg_voltage = 0.0f;
                 const bool mc_valid = (CANRX_get_signal(VEH, PM100DX_tractiveSystemVoltage, &ts_voltage) == CANRX_MESSAGE_VALID);
                 const bool chg_valid = (CANRX_get_signal(VEH, BRUSA513_dcBusVoltage, &chg_voltage) == CANRX_MESSAGE_VALID);
+                
+                if (SYS.contacts == SYS_CONTACTORS_PRECHARGE && !precharge_timer_started)
+                {
+                    drv_timer_init(&precharge_timer);
+                    drv_timer_start(&precharge_timer, 0);
+                    precharge_timer_started = true;
```
Comment (JoshLafleur) [inline components/bms_boss/src/BMS.c:151]: Shouldnt need this

File: components/bms_boss/src/BMS.c [hunk 6]
```diff
@@ -138,12 +143,26 @@ static void BMS100Hz_PRD(void)
                 float32_t chg_voltage = 0.0f;
                 const bool mc_valid = (CANRX_get_signal(VEH, PM100DX_tractiveSystemVoltage, &ts_voltage) == CANRX_MESSAGE_VALID);
                 const bool chg_valid = (CANRX_get_signal(VEH, BRUSA513_dcBusVoltage, &chg_voltage) == CANRX_MESSAGE_VALID);
+                
+                if (SYS.contacts == SYS_CONTACTORS_PRECHARGE && !precharge_timer_started)
```
Comment (JoshLafleur) [inline components/bms_boss/src/BMS.c:147]: You are already in a state where you are precharging

File: components/bms_boss/src/BMS.c [hunk 7]
```diff
@@ -138,12 +143,26 @@ static void BMS100Hz_PRD(void)
...
+                {
+                    drv_timer_init(&precharge_timer);
+                    drv_timer_start(&precharge_timer, 0);
+                    precharge_timer_started = true;
+                }
+                
                 if ((mc_valid || chg_valid) &&
-                    ((ts_voltage > 0.95f * BMS.pack_voltage) || (chg_voltage > 0.95f * BMS.pack_voltage)))
+                    ((ts_voltage > 0.95f * BMS.pack_voltage) || (chg_voltage > 0.95f * BMS.pack_voltage)) && 
+                (drv_timer_getElapsedTimeMs(&precharge_timer) >= PRECHARGE_MIN_TIME_MS))
```
Comment (JoshLafleur) [inline components/bms_boss/src/BMS.c:156]:
```suggestion
                (drv_timer_getState(&precharge_timer) == DRV_TIMER_EXPIRED))
```

File: components/bms_boss/src/BMS.c [hunk 8]
```diff
@@ -138,12 +143,26 @@ static void BMS100Hz_PRD(void)
...
                 {
                     SYS_SFT_cycleContacts();
+                    precharge_timer_started = false; //to reset for next time
```
Comment (JoshLafleur) [inline components/bms_boss/src/BMS.c:159]:
```suggestion
                    drv_timer_stop(&precharge_timer); //to reset for next time
```

File: components/bms_boss/src/Sys.c [hunk 9]
```diff
@@ -114,7 +114,7 @@ void SYS_SFT_openContactors(void)
     SYS.contacts = SYS_CONTACTORS_OPEN;
 }
 
-void SYS_SFT_cycleContacts(void)
+void closeSYS_SFT_cycleContacts(void)
```
Comment (JoshLafleur) [inline components/bms_boss/src/Sys.c:117]: Why did you change the name of this? This function does not close the contacts. It cycles the contacts state

File: components/bms_boss/src/BMS.c
Comment (JoshLafleur) [inline components/bms_boss/src/BMS.c:1]: Say you start precharging, and then while precharging you open the TSMS. The timer will not reset, so the next time you start precharging the timer will still be free running

File: components/bms_boss/src/BMS.c (hunk 3 above)
Comment (JoshLafleur) [reply components/bms_boss/src/BMS.c:150]: Also this timer will expire immediately. You are manually implementing things provided to you in the driver. This should be 
```suggestion
                    drv_timer_start(&precharge_timer, PRECHARGE_MIN_TIME_MS);
```

**Comment by JoshLafleur** [comment]
Also your test plan isnt descriptive enough. What are the steps to testing precharge termination? Can this be done on the testbench or does it need to be the vehicle? What are the edge cases that should be tested?

**Review by JoshLafleur — COMMENTED** [review]
You need a rebase onto master after your last merge. Otherwise lgtm

**Review by JoshLafleur — APPROVED** [review]
